crates/linker/src/pe.rs

```diff
@@ -0,0 +1,357 @@
+use std::{path::Path, time::Instant};
+
+use memmap2::Mmap;
+use object::{
+    pe::{ImageImportDescriptor, ImageNtHeaders64, ImageThunkData64},
+    read::pe::ImportTable,
+};
+use roc_collections::MutMap;
+use roc_error_macros::internal_error;
+
+#[allow(dead_code)]
+pub(crate) fn preprocess_windows(
+    exec_filename: &str,
+    _metadata_filename: &str,
+    _out_filename: &str,
+    _shared_lib: &Path,
+    _verbose: bool,
+    _time: bool,
+) -> object::read::Result<()> {
+    let total_start = Instant::now();
+    let exec_parsing_start = total_start;
+    let exec_file = std::fs::File::open(exec_filename).unwrap_or_else(|e| internal_error!("{}", e));
+    let exec_mmap = unsafe { Mmap::map(&exec_file).unwrap_or_else(|e| internal_error!("{}", e)) };
+    let exec_data = &*exec_mmap;
+    let _exec_obj = match object::read::pe::PeFile64::parse(exec_data) {
+        Ok(obj) => obj,
+        Err(err) => {
+            internal_error!("Failed to parse executable file: {}", err);
+        }
+    };
+
+    let _exec_parsing_duration = exec_parsing_start.elapsed();
+
+    let dynamic_relocations = DynamicRelocationsPe::new(exec_data);
+
+    println!("{:#x?}", dynamic_relocations);
+
+    Ok(())
+}
+
+#[derive(Debug)]
+struct DynamicRelocationsPe {
+    name_by_virtual_address: MutMap<u32, String>,
+
+    /// Map symbol name to the virtual address and file offset
+    /// where the actual address of this symbol is stored
+    address_and_offset: MutMap<String, (u32, u32)>,
+
+    /// Virtual offset of the first thunk
+    section_virtual_address: u32,
+
+    /// Offset in the file of the first thunk
+    section_offset_in_file: u32,
+
+    /// Offset in the file of the imports directory
+    #[allow(dead_code)]
+    imports_offset_in_file: u32,
+
+    /// The dummy .dll is the `dummy_import_index`th import of the host .exe
+    #[allow(dead_code)]
+    dummy_import_index: u32,
+}
+
+impl DynamicRelocationsPe {
+    fn new(data: &[u8]) -> Self {
+        Self::new_help(data).unwrap()
+    }
+
+    fn find_roc_dummy_dll(
+        import_table: &ImportTable,
+    ) -> object::read::Result<Option<(ImageImportDescriptor, u32)>> {
+        use object::LittleEndian as LE;
+
+        let mut index = 0;
+        let mut it = import_table.descriptors()?;
+        while let Some(descriptor) = it.next()? {
```

Comment:
Is there no `iter::find` or similar that could be used for this? Like enumerate to get the index and the find the specific descriptor? Just feels like an algorithm, but may not exist.